aten/src/ATen/native/EmbeddingBag.cpp

```diff
@@ -219,45 +245,52 @@ index_select_add(const Tensor &select_indices,
       offsets_include_last[offsets.numel()] = select_indices.numel();
       offsets_data = offsets_include_last.data();
     }
-
-#ifdef USE_FBGEMM
-    using float16 = uint16_t;
-    auto kernel_fp16_index_t = fbgemm_kernel_cache ?
-      fbgemm_kernel_cache->getCallback</* has_weight */ false, index_t, float16>(ddim) :
-      fbgemm::GenerateEmbeddingSpMDM<float16, index_t, index_t, float16>(
-        /* block_size */ddim,
-        /* has_weight */false,
-        /* normalize_by_lengths */false,
-        /* prefetch */16,
-        /* is_weight_positional */false,
-        /* use_offsets */true
-      );
-#else
+#if defined(USE_FBGEMM)
+    if (std::is_same<data_t, at::Half>::value) {
+      using float16 = uint16_t;
+      auto kernel_fp16_index_t = fbgemm_kernel_cache
+          ? fbgemm_kernel_cache
+                ->getCallback</* has_weight */ false, index_t, float16>(ddim)
+          : fbgemm::GenerateEmbeddingSpMDM<float16, index_t, index_t, float16>(
+                /* block_size */ ddim,
+                /* has_weight */ false,
+                /* normalize_by_lengths */ false,
+                /* prefetch */ 16,
+                /* is_weight_positional */ false,
+                /* use_offsets */ true);
+      at::parallel_for(
+          0, output_size, 1, [&](index_t start_idx, index_t end_idx) {
+            bool success = kernel_fp16_index_t(
+                /* output_size */ end_idx - start_idx,
+                /* index_size */ offsets_data[end_idx] -
+                    offsets_data[start_idx],
+                /* data_size */ src.size(0),
+                /* input */ reinterpret_cast<const float16*>(src_data),
+                /* indices */ select_indices_data + offsets_data[start_idx],
+                /* offsets_or_lengths */ offsets_data + start_idx,
+                /* weights */ nullptr,
+                /* output */
+                reinterpret_cast<float16*>(output_data + start_idx * ddim));
+            if (!success) {
+              fbgemm_spmdm_report_error_(
+                  end_idx - start_idx,
+                  offsets_data[end_idx] - offsets_data[start_idx],
+                  src.size(0),
+                  offsets_data + start_idx,
+                  select_indices_data + offsets_data[start_idx]);
+            }
+          });
+      return;
+    }
+#endif
+#if !defined(USE_FBGEMM) or defined(BUILD_CAFFE2)
```

Comment:
Do you need to check the data type is bfloat16 here?

Reply:
> Do you need to check the data type is bfloat16 here?

Caffe2 support both fp16/bf16, so we do not need to check bf16 here. The difference is we can convert bf16->fp32 with vec but convert fp16->fp32 with scalar.

Reply:
But the code below explicitly uses BFloat16. Or, did I miss anything here? Also, there is also vec support for fp16<->fp32. This can be further improved in the future if we are using scalar conversion currently.

Reply:
Only these two `using`s are explicitly related to bf16 but not check the data type. I think it's unnecessary.
```
    using bVec = vec::Vectorized<BFloat16>;
    using fVec = vec::Vectorized<float>;
```
Caffe2 kernels are same for bf16/fp16.
And fp32->bf16/fp16 part are different, I check the data type there.

For vec support fp16<->fp32 on CPU, I do not find them on cpu/vec folders. I will do some investigate in the future.